### ISSUE TITLE
coordinate efforts on Go tooling for protolock

### ISSUE THREAD

**Issue by nilslice** [spotify/protoman#33]
Hey - it was brought to my attention that there may be some efforts to port `protoman` backward compatibility tooling to Go, and thought I would share a tool the community has helped improved significantly since my initial release: [`protolock`](https://github.com/nilslice/protolock)

There have been a number of community-driven PRs and usage from within orgs including Fanatics, Lyft (internal and Envoy), Salesforce, Google (Istio), and more. If Spotify were interested in adopting the tool as well, I'm sure we can plan to incorporate features `protoman` supports that `protolock` does not yet. We recently added support for plugins (a la `protoc`-style executables), which can be written in any language and extend the base rules baked into `protolock`.  

If you have already made progress elsewhere, ignore this and carry on :) -- just figured I'd share another option with hopes of bringing more skilled contributors together. 

Thanks!

**Comment by jhaals** [comment]
Hey @nilslice, first of all thanks for reaching out. We are really interested in improving the tooling for proto management and validation. This is the first time I hear about protolock and it looks cool :) 

Our public protoman in it's current form has been put on hold. We have however rewritten and simplified it to just a CLI in Golang which just does fetch in it's current form. The registry is a git repo with all our protos, protoman is using the github API to fetch the specified package and it's depenedencies. We are using prototool by lyft as a build step to do validation in our shared proto repo when users publish new changes.
We would like to incorporate some of our own lint checks and it would be awesome to put everything into one tool that can be extended. Since the current version of the protos are stored in git we are planning to rely on git comparison between master and the current pr/branch to calculate the state changes instead of having to commit an additional lock file. Do you think that could be something that could be incorporated into protolock as a new/secondary way of comparing the state? 

I will bring this up with my team and also see if we can upload our current iteration of protoman to github to give some more context on how we currently use our tool.

Thanks again for reaching out!

**Comment by nilslice** [comment]
Thanks for the reply, and I apologize for my delayed response!

> Since the current version of the protos are stored in git we are planning to rely on git comparison between master and the current pr/branch to calculate the state changes instead of having to commit an additional lock file.

I considered using Git as a tool to track and source proto versions as well, but have concluded that the lock file is more efficient and less restrictive. Otherwise, using `git` is a strict requirement (some users have other VCS), and using `protoc` is a requirement, since generating the descriptor set must be done before the comparison. Using `protoc` is not as big of a deal, as we are working with protobufs after all :)  

Avoiding the use of a lock file is completely understandable, but I've found it is actually quite nice to have a readable, tracked reference to the state of your protos.

> Do you think that could be something that could be incorporated into protolock as a new/secondary way of comparing the state?

Personally, I find the use of the descriptor set a bit more cumbersome to work with when comparing the states. I'm looking forward to seeing any real code that does the comparisons, and code to provide extensible plugins for adding custom checks/rules - to my knowledge, this has not started or reached a state where it could be released. Although I'd like to say it could be supported in `protolock`, I have no plans to start that work and add `git` and `protoc` as dependencies. I think the right path is a standalone tool that covers a smaller set of responsibilities.

I would be really interested to see your current iteration, but understand some things need to stay internal :)
